### PR TITLE
Refactor session ↔︎ leader interaction for durable shared subscriptions 

### PR DESCRIPTION
**Pull request by savonarola**
Fixes [EMQX-13515](https://emqx.atlassian.net/browse/EMQX-13515)

Release version: v/e5.9.0

Here, we make several _internal_ model simplifications, resulting in less code, less complex code, and thus, hopefully, in more robust code. 

These changes _do not affect_ the code interacting with durable shared subscriptions, e.g., durable session.

##  Old code structure

* DS session delegates shared sub handling to `emqx_persistent_session_ds_shared_subs`, symmetrically to `emqx_persistent_session_ds_subs`.
* `emqx_persistent_session_ds_shared_subs` talks to `emqx_ds_shared_sub_agent` hiding session internals from it.
  * Also, delegates subscription/unsubscription requests to `emqx_ds_shared_sub_agent`; 
  * receives events from `emqx_ds_shared_sub_agent`: stream leases and revokes;
  * sends stream progress to `emqx_ds_shared_sub_agent`.
* `emqx_ds_shared_sub_agent` belongs to `emqx_ds_shared_sub` up and talks to shared sub leaders. Technically it does little itself but delegates calls to a collection of `emqx_ds_shared_sub_group_sm`'s instances, one for each topic subscription.
* `emqx_ds_shared_sub_group_sm` represents a stateful connection to a leader. It actually talks to the leader, receives streams from it, etc.
* `emqx_ds_shared_sub_leader` is responsible for renewing and distributing streams across connected  `emqx_ds_shared_sub_group_sm`s and recording progress. 
 
## Changes

The general structure or layers remains the same. 

### `emqx_persistent_session_ds_shared_subs`

We radically simplify `emqx_persistent_session_ds_shared_subs`. (The first step was done in  https://github.com/emqx/emqx/pull/14227). 
* As with regular subs, we identify subscription not by topic but by a unique ID. So When a session resubscribes to a topic, this is a completely new subscription.
* We remove stream "finalization" after unsubscription from it.

The relevant commit is https://github.com/emqx/emqx/pull/14428/commits/f65eccd5d298ca386f9e58bf964f687472805900

Now `emqx_persistent_session_ds_shared_subs` does not have any additional state except `emqx_ds_shared_sub_agent` to which it talks to receive streams.

### `emqx_ds_shared_sub_agent` 

* Here we remove unnecessary preliminary unpacking of messages sent to `subscribers` (former `group_sm`s). This is not necessary for dispatching messages to individual subscription handlers.
* We address individual subscription handlers by an additional ref to guarantee the abscence of any stale messages.
* We also provide to individual subscription handlers (former `group_sm`s) an ability to request a complete recreation (in case inconsistent state is detected).

The relevant commit is https://github.com/emqx/emqx/pull/14428/commits/81271d3b68d0335a1ffc95754c4f534f315769ba

### `emqx_ds_shared_sub_group_sm`

* We rename "Group State Machine" handler module `emqx_ds_shared_sub_group_sm` to `emqx_ds_shared_sub_subscriber`. We identify it throughout the code as `ssubscriber`s, _shared subscribers_.
* It does not have cyclic state changes and a complex state machine anymore. On invalidation, `ssubscriber` finishes and requests the agent to be recreated. The states are [new] -> `connecting` (to the leader) -> `connected` -> `unsubscribing` -> [deleted].
* We discard "versioned sets of streams" (Andrew's idea). Instead, we grant and revoke individual streams. This decouples the `ssubscriber`'s states from stream states, and the related events are handled mostly independently.

The corresponding commit is https://github.com/emqx/emqx/pull/14428/commits/12acd63df2c97b3f6bfd928c050a895428bc0674.

### `emqx_ds_shared_sub_leader`

* We mostly keep the same the logic related to stream renewal and persistence.
* We go away from the outdated "agent" naming. Previously, a session could be connected to a leader only once through its agent. Now, a session may have several ssubscribers connected (1 active and several unsubscribing), so this naming became more misleading.
* We convert the module to a `gen_server` since it always appeared to have a single state.
* Reflecting the new `emqx_ds_shared_sub_subscriber` updates, we change the model. Instead of having subscribers with versioned sets of revoking/granted streams, we have a plain structure of streams with status and owner. 
* We try to keep subscriber connection/validity handling distinct from stream granting/revocation. This results in smaller and less nested handlers.

The corresponding commit is https://github.com/emqx/emqx/pull/14428/commits/397517883119595e6d05e2181ee0ac7d1854d76e

I tried to group the described changes into distinct relevant commits. However, since the changes were quite significant, individual commits do not keep the code in a working state. There are some final changes to make everything up and running.

## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked:

- [x] Added tests for the changes (tests present, we changed the internal behaviour)
- [na] Added property-based tests for code which performs user input validation
- [na] Change log has been added to `changes/(ce|ee)/(feat|perf|fix|breaking)-<PR-id>.en.md` files (internal change of unreleased feature).
- [x] For internal contributor: there is a jira ticket to track this change
- [na] Schema changes are backward compatible (not compatible but only for hidden parameters of an unreleased feature).

[EMQX-13515]: https://emqx.atlassian.net/browse/EMQX-13515?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apps/emqx_ds_shared_sub/src/emqx_ds_shared_sub_leader.erl [hunk 1]
```diff
@@ -1014,59 +769,87 @@ shuffle(L0) ->
...
+                Stream,
+                #{status := ?stream_granted, ssubscriber_id := SSubscriberId},
+                SSubscribersAcc
+            ) ->
+                case SSubscribersAcc of
+                    #{SSubscriberId := Streams} ->
+                        SSubscribersAcc#{SSubscriberId => [Stream | Streams]};
+                    _ ->
+                        SSubscribersAcc
+                end;
```
Comment (thalesmg) [inline apps/emqx_ds_shared_sub/src/emqx_ds_shared_sub_leader.erl:840]: Micro nit:

```suggestion
                emqx_utils_maps:update_if_present(
                    SSubscriberId,
                    fun(Streams) -> [Stream | Streams] end,
                    SSubscribersAcc
                );
```

File: apps/emqx_ds_shared_sub/src/emqx_ds_shared_sub_subscriber.erl [hunk 2]
```diff
@@ -0,0 +1,473 @@
...
+-type t() :: #{
+    session_id := emqx_persistent_session_ds:id(),
+    share_topic_filter := emqx_persistent_session_ds:share_topic_filter(),
+    id := emqx_persistent_session_ds:id(),
+    send_after := fun((non_neg_integer(), term()) -> reference()),
+
+    status := status(),
+    leader => emqx_ds_shared_sub_proto:leader(),
+    streams => #{emqx_ds:stream() => stream_data()},
+    timers => #{
```
Comment (thalesmg) [inline apps/emqx_ds_shared_sub/src/emqx_ds_shared_sub_subscriber.erl:109]: Nit, apparently this is:

```suggestion
    streams := #{emqx_ds:stream() => stream_data()},
    timers := #{
```

**Review by thalesmg — APPROVED** [review]
Looks good, but a review from someone more familiar with this would be best.  

:beers:

File: apps/emqx_ds_shared_sub/src/emqx_ds_shared_sub_registry.erl [hunk 3]
```diff
@@ -41,13 +43,14 @@ child_spec() ->
...
+    emqx_ds_shared_sub_proto:ssubscriber_id(),
     emqx_persistent_session_ds:share_topic_filter()
 ) -> ok.
-leader_wanted(Agent, AgentMetadata, ShareTopic) ->
+leader_wanted(SSubscriberId, ShareTopic) ->
     {ok, Pid} = ensure_elector_started(ShareTopic),
-    emqx_ds_shared_sub_proto:agent_connect_leader(Pid, Agent, AgentMetadata, ShareTopic).
+    emqx_ds_shared_sub_proto:send_to_leader(Pid, ?ssubscriber_connect(SSubscriberId, ShareTopic)).
+
+% ShareTopic, ?ssubscriber_connect_match(SSubscriberId)).
```
[the remark this answers is not in the document]
Comment (savonarola) [reply apps/emqx_ds_shared_sub/src/emqx_ds_shared_sub_registry.erl]: ✅

**Review by thalesmg — APPROVED** [review]

File: apps/emqx_ds_shared_sub/src/emqx_ds_shared_sub_subscriber.erl [hunk 4]
```diff
@@ -0,0 +1,473 @@
...
+%% Possible stream transitions are as follows:
+%%
+%% [no stream] -> stream_granted -> stream_revoking -> [no stream]
+
+-type stream_status() :: ?stream_granted | ?stream_revoking.
+
+-type stream_data() :: #{
+    status := stream_status(),
+    progress := progress(),
+    use_finished := boolean()
```
Comment (keynslug) [inline apps/emqx_ds_shared_sub/src/emqx_ds_shared_sub_subscriber.erl:90]: Nit. I think a comment describing what it is and what it is for would be welcome here.

File: apps/emqx_ds_shared_sub/src/emqx_ds_shared_sub_subscriber.erl [hunk 5]
```diff
@@ -0,0 +1,473 @@
...
+        end,
+    St1 = St0#{streams => Streams1},
+    ok = notify_progress(St1, Stream),
+    {ok, Events, St1};
+%%
+%% Revoke stream
+%%
+on_info(St, ?leader_revoke_match(_Leader, _Stream)) when ?is_connecting(St) ->
+    %% Should never happen.
+    reset(St);
```
Comment (keynslug) [inline apps/emqx_ds_shared_sub/src/emqx_ds_shared_sub_subscriber.erl:232]: Nit. Could be great to have better visibility into such events. Or perhaps something like a test-profile-only assertion?

**Review by keynslug — COMMENTED** [review]
Pretty solid refactoring. Definitely became easier to comprehend and navigate.

File: apps/emqx_ds_shared_sub/src/emqx_ds_shared_sub_leader.erl [hunk 6]
```diff
@@ -63,169 +74,131 @@
...
-) ->
-    Data1 = with_agent(Data0, Agent, fun() ->
-        disconnect_agent(Data0, Agent, StreamProgresses, Version)
+    {noreply, St1};
+handle_info(?ssubscriber_revoke_finished_match(SSubscriberId, Stream), St0) ->
+    St1 = with_valid_ssubscriber(St0, SSubscriberId, fun() ->
+        handle_revoke_finished(St0, SSubscriberId, Stream)
     end),
-    {keep_state, Data1};
+    {noreply, St1};
```
[the remark this answers is not in the document]
Comment (savonarola) [reply apps/emqx_ds_shared_sub/src/emqx_ds_shared_sub_leader.erl:183]: It is considered OK if a leader (re)sends `revoked` and the ssubscriber already has no data related to the stream. So there may be no progress to send.

File: apps/emqx_ds_shared_sub/src/emqx_ds_shared_sub_agent.erl [hunk 7]
```diff
@@ -14,44 +17,53 @@
...
 -type subscription() :: emqx_persistent_session_ds_shared_subs_agent:subscription().
 -type share_topic_filter() :: emqx_persistent_session_ds:share_topic_filter().
 -type subscription_id() :: emqx_persistent_session_ds_shared_subs_agent:subscription_id().
--type group_id() :: subscription_id().
 
 -type options() :: #{
     session_id := emqx_persistent_session_ds:id()
 }.
 
+-record(ssubscriber_entry, {
```
[the remark this answers is not in the document]
Comment (savonarola) [reply apps/emqx_ds_shared_sub/src/emqx_ds_shared_sub_agent.erl:45]: Do you refer to the "entry" data item or the ssubscriber terminology in general?

**Comment by savonarola** [comment]
I will add some updates in a follow-up.

**Comment by emqxqa** [comment]
[TestExecution](https://emqx.atlassian.net/browse/QA-6700)

**Pull request merged**